cli/js/testing.ts

```diff
@@ -95,41 +95,167 @@ interface TestCase {
+  result: TestResult;
+}
+
+interface EndMsg {
+  kind: MsgKind.End;
+  stats: TestStats;
+}
+
+type RunTestsMessage = StartMsg | TestMsg | EndMsg;
+
+class RunTestsIterable implements AsyncIterableIterator<RunTestsMessage> {
+  readonly testsToRun: TestDefinition[];
+  readonly testResults: TestResult[];
+
+  private hasStarted = false;
+  private hasFinished = false;
+  private currentTestIndex = 0;
+
+  failFast: boolean;
 
-  const stats: TestStats = {
+  stats: TestStats = {
     measured: 0,
     ignored: 0,
     filtered: 0,
     passed: 0,
     failed: 0
   };
 
-  const testCases = testsToRun.map(
-    ({ name, fn }): TestCase => {
+  constructor(
+    tests: TestDefinition[],
+    only: RegExp,
+    skip: RegExp,
+    failFast: boolean
+  ) {
+    this.testsToRun = tests.filter(
+      ({ name }): boolean => only.test(name) && !skip.test(name)
+    );
+    this.failFast = failFast;
+    this.testResults = this.testsToRun.map(
+      ({ name, fn }): TestResult => {
+        return {
+          name,
+          fn,
+          duration: 0,
+          hasRun: false,
+          failed: false,
+          error: undefined
+        };
+      }
+    );
+  }
+
+  async next(): Promise<IteratorResult<RunTestsMessage>> {
+    if (!this.hasStarted) {
+      this.hasStarted = true;
       return {
-        name,
-        fn,
-        timeElapsed: 0,
-        error: undefined
+        done: false,
+        value: {
+          kind: MsgKind.Start,
+          tests: this.testsToRun.length,
+          stats: this.stats
+        }
       };
     }
-  );
+
+    if (this.hasFinished) {
+      return {
+        done: true,
+        value: {
+          kind: MsgKind.End,
+          stats: this.stats
+        }
+      };
+    }
+
+    const testResult = this.testResults[this.currentTestIndex];
+    if (testResult === undefined) {
+      throw new Error(`Missing test case: ${this.currentTestIndex}`);
+    }
+
+    if (testResult.hasRun) {
+      throw new Error(`Trying to run test case again: ${testResult.name}`);
+    }
+
+    // TODO(bartlomieju): using permformance here doesn't make sense, as there
+    // is no guarantee that we're running with `--allow-hr` perm
+    try {
+      const start = performance.now();
+      await testResult.fn();
+      const end = performance.now();
+      testResult.duration = end - start;
+      this.stats.passed++;
+    } catch (err) {
+      testResult.failed = true;
+      testResult.error = err;
+      this.stats.failed++;
+      if (this.failFast) {
+        this.hasFinished = true;
+      }
+    } finally {
+      testResult.hasRun = true;
+    }
+
+    this.currentTestIndex++;
+    if (this.currentTestIndex === this.testResults.length) {
+      this.hasFinished = true;
+    }
+
+    return {
+      done: false,
+      value: {
+        kind: MsgKind.Test,
+        result: testResult
+      }
+    };
+  }
+
+  [Symbol.asyncIterator](): AsyncIterableIterator<RunTestsMessage> {
+    return this;
+  }
```

Comment:
Can we use an async generator instead of a `next()` method?

Reply:
I'm open to that, is there an advantage doing that?

Reply:
Then no ugly `next()` method is exposed. I've been meaning to get rid of the one in `Listener`.

Just realised this class is private, though. Who cares.

Reply:
Still should be consistent... would rather have async generator methods.  I don't know if it empowers V8 to optimise stuff behind the scenes or not, but it _could_.  If you implement the shape of a generator, it may limit V8s ability to optimise.